go/ct/gen/block_context.go

```diff
@@ -68,13 +232,211 @@ func (*BlockContextGenerator) Generate(rnd *rand.Rand, revision common.Revision)
+			} else {
+				clauses = append(clauses, fmt.Sprintf(
+					"%v ∉ [BlockNumber-256..BlockNumber-1]",
+					variable,
+				))
+			}
+		}
+	}
+
+	if b.valueConstraint != nil {
+		for variable, value := range b.valueConstraint {
+			op := "+"
+			if value > 0 {
+				op = "-"
+			} else {
+				value = -value
+			}
+			clauses = append(clauses, fmt.Sprintf(
+				"%v = BlockNumber%s%d",
+				variable,
+				op,
+				value,
+			))
+		}
+	}
+
+	sort.Slice(clauses, func(i, j int) bool {
+		return clauses[i] < clauses[j]
+	})
+
+	return strings.Join(clauses, " Λ ")
+}
+
+// RestricVariableToOneOfTheLast256Blocks adds a constraint on the variable
+// so that this generator assigns a value to it referencing one of the last 256 blocks.
+func (b *BlockContextGenerator) RestrictVariableToOneOfTheLast256Blocks(variable Variable) {
+	if b.unsatisfiable {
+		return
+	}
+	if b.rangeConstraints == nil {
+		b.rangeConstraints = make(map[Variable]bool)
+	}
+	if inRange, ok := b.rangeConstraints[variable]; ok {
+		if !inRange {
+			b.markUnsatisfiable()
+		}
+	} else {
+		b.rangeConstraints[variable] = true
+	}
+}
+
+// RestrictVariableToNoneOfTheLast256Blocks adds a constraint on the variable
+// so that this generator assigns a value to it referencing none of the last 256 blocks.
+func (b *BlockContextGenerator) RestrictVariableToNoneOfTheLast256Blocks(variable Variable) {
+	if b.unsatisfiable {
+		return
+	}
+	if b.rangeConstraints == nil {
+		b.rangeConstraints = make(map[Variable]bool)
+	}
+	if inRange, ok := b.rangeConstraints[variable]; ok {
+		if inRange {
+			b.markUnsatisfiable()
+		}
+	} else {
+		b.rangeConstraints[variable] = false
+	}
+}
+
+// SetBlockNumberOffsetValue adds a constraint on the variable so that it is
+// assigned with a value that is offseted from the block number.
+func (b *BlockContextGenerator) SetBlockNumberOffsetValue(variable Variable, value int64) {
+	if b.unsatisfiable {
+		return
+	}
+	if b.valueConstraint == nil {
+		b.valueConstraint = make(map[Variable]int64)
+	}
+	if existingValue, ok := b.valueConstraint[variable]; ok {
+		if existingValue != value {
+			b.markUnsatisfiable()
+		}
+	} else {
+		b.valueConstraint[variable] = value
+	}
+}
+
+// SetRevision adds a constraint on the State's revision.
+func (b *BlockContextGenerator) SetRevision(revision Revision) {
+	b.AddRevisionBounds(revision, revision)
+}
+
+// AddRevisionBounds adds a constraint on the State's revision.
+func (b *BlockContextGenerator) AddRevisionBounds(lower, upper Revision) {
+	if b.unsatisfiable {
+		return
+	}
+	if lower > upper || lower < 0 || upper < 0 {
+		b.markUnsatisfiable()
+		return
+	}
+	min, err := GetForkBlock(lower)
+	if err != nil {
+		b.markUnsatisfiable()
+		return
+	}
+	max, err := GetForkBlock(upper)
+	if err != nil {
+		b.markUnsatisfiable()
+		return
+	}
+	len, err := GetBlockRangeLengthFor(upper)
+	if err != nil {
+		b.markUnsatisfiable()
+		return
+	}
+	if len == math.MaxUint64 {
+		max = math.MaxUint64
+	} else {
+		max += len - 1
```

Comment:
nit: the overflow here may happen also for values of `len` smaller than `math.MaxUint64`